Create rule S6367: Thread local variables should not be used in coroutines (#479)
SonarSource/rspec: rules/S6367/cfamily/rule.adoc

> In contrast to normal function, the coroutines can suspend and later resume their execution. Depending on the program, the coroutine may resume on a different thread of execution, compared to one it was started or run previously.

Reviewer: ```suggestion
In contrast to normal functions, coroutines can suspend and later resume their execution. Depending on the program, the coroutine may resume on a different thread of execution than the one it was started or run previously on.
```

> Due above, the access to the "same" variable with `thread_local` storage, may produce different values, as illustrated below:

Reviewer: ```suggestion
Therefore, the access to the "same" variable with `thread_local` storage may produce different values as illustrated below:
```

----
thread_local std::vector<Task> tasks;
lazy<Thingy> doSomething() {

>  // evalutation started on thread t1, one function was invoked

Reviewer: I don't understand the second part of the sentence "one function was invoked". What do you mean by that?

Reviewer 2: it should be "once function was invoked", but I think can safely remove it.

 /* .... */

>  std::size_t const taskCount = tasks.size(); // value specific to thread t1

Reviewer: ```suggestion
const std::size_t taskCount = tasks.size(); // value specific to thread t1
```
I'm all for east const but I'm not sure it's consistent with other examples.

>  auto result = co_await produceThingy(); // suspend execution, resumes it on thread t2

Reviewer: ```suggestion
 auto result = co_await produceThingy();
 // after co_await, execution resumes on thread t2
```

 for (std::size_t i = 0; i < taskCount; ++i) {
    tasks[i].evalute(); // access value specific to t2

>     // miss some threads if t1:taskCount.size() < t2:taskCount.size()

Reviewer: ```suggestion
    // miss some tasks if t1:taskCount.size() < t2:taskCount.size()
```

    // undefined behavior if t1:taskCount.size() > t2:taskCount.size()

>  }  

Reviewer: The indentation is slightly off.

  return result;
}
----
This behavior is surprising and unintuitive in comparison to normal functions that are always evaluated on a single thread.

> The same issue can happen for the use of different thread-local variables if their values are connected.

Reviewer: I'm not sure what you mean by "if their values are connected".

Reviewer 2: Added example

> Moreover, access to `thread-local` variable defined inside the coroutine, may read uninitialized memory. 

Reviewer: ```suggestion
Moreover, access to `thread-local` variables defined inside the coroutine may read uninitialized memory. 
```

Each such variable is initialized when a specific thread enters the function for the first time, 

> and if the function was never called from a thread on which the coroutine was resumed, it is uninitialized.

Reviewer: ```suggestion
and if the function was never called from a thread on which the coroutine is resumed, it is uninitialized.
```

> This rule raises an issue on the declaration of `thread_local` and access to  `thread_local` variables

Reviewer: ```suggestion
This rule raises an issue on the declaration of `thread_local` variables and access to `thread_local` variables
```

> in the coroutine body.

Reviewer: ```suggestion
in coroutines.
```

== Noncompliant Code Example

----
thread_local std::vector<Task> tasks;
lazy<Thingy> doSomething() {
  thread_local Tasks localTask; // Noncompliant

>   std::size_t const taskCount = tasks.size(); // Noncompliant

Reviewer: ```suggestion
  const std::size_t taskCount = tasks.size(); // Noncompliant
```
ditto.

  }  
  co_return result;
}
----
This behavior is surprising and unintuitive compared to normal functions that are always evaluated on a single thread.
The same issue can happen for the use of different thread-local variables if their values are interconnected (e.g., one is the address of the buffer, and the other is the number of elements in the buffer).

Moreover, access to `thread-local` variables defined inside the coroutine may read uninitialized memory. 
Each such variable is initialized when a specific thread enters the function for the first time, 
and if the function was never called from a thread on which the coroutine is resumed, it is uninitialized.

This rule raises an issue on the declaration of `thread_local` variables and access to `thread_local` variables
in coroutines.

== Noncompliant Code Example

----
thread_local std::vector<Decorator> decorators;
lazy<Thingy> doSomething() {
  thread_local Decorator localDecorator; // Noncompliant
  const std::size_t decoratorCount = decorators.size(); // Noncompliant
  /* ... */
  auto result = co_await produceThingy();
  for (std::size_t i = 0; i < taskCount; ++i) {
    decorators[i].modify(result);
  }  

>   localDecorator..modify(result); // Noncompliant

Reviewer: ```suggestion
  localDecorator.modify(result); // Noncompliant
```

In contrast to normal functions, coroutines can suspend and later resume their execution. Depending on the program, the coroutine may resume on a different thread of execution than the one it was started or run previously on.

Therefore, the access to the "same" variable with `thread_local` storage may produce different values as illustrated below:
----
thread_local std::vector<Task> tasks;
lazy<Thingy> doSomething() {
 // evalutation started on thread t1
 /* .... */
const std::size_t taskCount = tasks.size(); // value specific to thread t1
 auto result = co_await produceThingy();
 // after co_await, execution resumes on thread t2
 for (std::size_t i = 0; i < taskCount; ++i) {
   tasks[i].evalute(); // access value specific to t2
   // miss some tasks if t1:taskCount.size() < t2:taskCount.size()
   // undefined behavior if t1:taskCount.size() > t2:taskCount.size()
 }  
  return result;

> }

Reviewer: The indentation is off around here.

>   return result;

Reviewer 3: As I understood from https://en.cppreference.com/w/cpp/language/coroutines a coroutine cannot use plain `return`. Should it be `co_return`?
```suggestion
  co_return result;
```
Same for the noncompliant example

Reviewer 2: Yes, you are right updated in both places.

> thread_local std::vector<Task> tasks;
> lazy<Thingy> doSomething() {
>   // evalutation started on thread t1
>   /* .... */
>   const std::size_t taskCount = tasks.size(); // value specific to thread t1
>   auto result = co_await produceThingy();
>   // after co_await, execution resumes on thread t2
>   for (std::size_t i = 0; i < taskCount; ++i) {
>     tasks[i].evalute(); // access value specific to t2
>     // miss some tasks if t1:taskCount.size() < t2:taskCount.size()
>     // undefined behavior if t1:taskCount.size() > t2:taskCount.size()
>   }  
>   co_return result;
> }

Reviewer 3: As I understood, "task" in this example is unrelated to coroutines, it is just a classic chunk of computation. It might be confusing, since "task" is often used to name a promise object for some coroutines. I propose using a different term, less associated with coroutines. For example:
```suggestion
thread_local std::vector<Formula> formulae;
lazy<Thingy> doSomething() {
  // evaluation started on thread t1
  /* .... */
  const std::size_t formulaCount = formulae.size(); // value specific to thread t1
  auto result = co_await produceThingy();
  // after co_await, execution resumes on thread t2
  for (std::size_t i = 0; i < formulaCount; ++i) {
    formulae[i].evalute(); // access value specific to t2
    // miss some formulae if t1:formulaCount < t2:formulaCount
    // undefined behavior if t1:formulaCount > t2:formulaCount
  }  
  co_return result;
}
```
Or even something as banal as:
```suggestion
thread_local std::vector<int> nums;
lazy<Thingy> doSomething() {
  // evaluation started on thread t1
  /* .... */
  const std::size_t numCount = nums.size(); // value specific to thread t1
  auto result = co_await produceThingy();
  // after co_await, execution resumes on thread t2
  for (std::size_t i = 0; i < numCount; ++i) {
    nums[i] += i; // access value specific to t2
    // miss some tasks if t1:numCount < t2:numCount
    // undefined behavior if t1:numCount > t2:numCount
  }  
  co_return result;
}
```
Same comment applies for the noncompliant example.

Reviewer 3: On an unrelated note, a few more misprints in the example:
```suggestion
thread_local std::vector<Task> tasks;
lazy<Thingy> doSomething() {
  // evaluation started on thread t1
  /* .... */
  const std::size_t taskCount = tasks.size(); // value specific to thread t1
  auto result = co_await produceThingy();
  // after co_await, execution resumes on thread t2
  for (std::size_t i = 0; i < taskCount; ++i) {
    tasks[i].evalute(); // access value specific to t2
    // miss some tasks if t1:taskCount < t2:taskCount
    // undefined behavior if t1:taskCount > t2:taskCount
  }  
  co_return result;
}
```

Reviewer 2: Indeed the use of Task may be misleading, updated the example and fixed the code.

----
This behavior is surprising and unintuitive in comparison to normal functions that are always evaluated on a single thread.
The same issue can happen for the use of different thread-local variables if their values are interconnected

> (e.g. one is address of the buffer, and other number of elements in it).

Reviewer: ```suggestion
(e.g. one is the address of the buffer, and the other is the number of elements in the buffer).
```

> This behavior is surprising and unintuitive in comparison to normal functions that are always evaluated on a single thread.
> The same issue can happen for the use of different thread-local variables if their values are interconnected
> (e.g. one is the address of the buffer, and the other is the number of elements in the buffer).

Reviewer 4: ```suggestion
This behavior is surprising and unintuitive compared to normal functions that are always evaluated on a single thread.
The same issue can happen for the use of different thread-local variables if their values are interconnected (e.g., one is the address of the buffer, and the other is the number of elements in the buffer).
```

Moreover, access to `thread-local` variables defined inside the coroutine may read uninitialized memory. 
Each such variable is initialized when a specific thread enters the function for the first time, 
and if the function was never called from a thread on which the coroutine is resumed, it is uninitialized.

This rule raises an issue on the declaration of `thread_local` variables and access to `thread_local` variables
in coroutines.

== Noncompliant Code Example

----
thread_local std::vector<Task> tasks;
lazy<Thingy> doSomething() {

>   thread_local Tasks localTask; // Noncompliant

Reviewer 3: ```suggestion
  thread_local Task localTask; // Noncompliant
```

Reviewer 2: Fixed.
